ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,330 @@
+        super().__init__(**kwargs)
+        
+        self.all_loads = [self.load_dl1_images,
+                          self.load_dl1_parameters,
+                          self.load_dl2_geometry,
+                          self.load_simulated,
+                          self.load_true_images,
+                          self.load_trigger,
+                          self.load_instrument]
+        
+        self.subarray = SubarrayDescription.from_hdf(self.input_url)
+        self.h5file = tables.open_file(self.input_url, mode="r")
+
+        try:
+            self.structure = get_structure(self.h5file)
+        except ValueError:
+            self.structure = None
+
+        self.trigger_table = read_table(self.h5file, TRIGGER_TABLE)
+
+        self.instrument_table = None
+        if self.load_instrument:
+            table = self.subarray.to_table()
+            optics = self.subarray.to_table(kind="optics")
+            optics["optics_index"] = np.arange(len(optics))
+            optics.remove_columns(["name", "description", "type"])
+            table = join(
+                table,
+                optics,
+                keys="optics_index",
+                # conflicts for TAB_VER, TAB_TYPE, not needed here, ignore
+                metadata_conflicts="silent",
+            )
+
+            table.remove_columns(["optics_index", "camera_index"])
+            self.instrument_table = table
+
+        self.shower_table = None
+        if self.load_simulated and SHOWER_TABLE in self.h5file:
+            self.shower_table = read_table(self.h5file, SHOWER_TABLE)
+
+    def close(self):
+        self.h5file.close()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _get_reconstructors(self):
+
+        if GEOMETRY_GROUP in self.h5file:
+            g = self.h5file.root[GEOMETRY_GROUP]
+        else:
+            raise ValueError(f"No DL2 geometry data in h5file: {self.h5file}")
+        
+        reconstructors = g._v_children.keys()
+        
+        return reconstructors
+
+    def _read_telescope_table(self, group, tel_id):
+
+        if self.structure == "by_id":
+            key = f"{group}/tel_{tel_id:03d}"
+            condition = None
+        else:
+            key = f"{group}/{self.subarray.tel[tel_id]!s}"
+            condition = f"tel_id == {tel_id}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=condition)
+        else:
+            table = Table()
+
+        return table
+    
+    def _read_telescope_DL2_geometry(self, group, reconstructor):
+        
+        key = f"{group}/{reconstructor}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=None)
+        else:
+            table = Table()
+
+        return table
+
+    def read_subarray_events(self, labels=None, sort_by=["obs_id", "event_id"]):
```

Comment:
The return type of this function depends on the structure of the file. I think this is not good, since then the code needs to know it. I implemented the functions so they give a well defined output independent of the structure in the file, so that code can be agnostic about that. I think this is an important property we should keep!

Reply:
Not sure if I get what is the problem here (which should be easily fixable anyway, since we have all the tools here).

I don't understand "since then the code needs to know it."
The code will understand via the conditions what to do and will return what I will write in the docstring for the user (sorry I didn't update it)

Reply:
the point is exactly to make the code check the dataset split mode so I don't have to worry about it and with 1 function I can ask for anything (available) regardless of the input file nature

Reply:
I am talking about the code calling this one. The structure of the return value of this function depends on the structure in the file, correct?

Reply:
not really, you can read as tel_types from a tel_id file as well as tel_ids from a tel_type file

that is why I did it like this: the return value depends only on the labels (tel_ids or tel_type) that the user asks for.
The usage is completely blind to the dataset split (as it should since it doesn't change the contents that you read)

Reply:
reverted